### PR TITLE
feat: precompile JSX to string transform

### PR DESCRIPTION
**Pull request by bartlomieju**
This PR introduces a new JSX transform that is optimized for server-side rendering. It works by serializing the static parts of a JSX template into static string arrays at compile time. Common JSX templates render around 200-500 nodes in total and the transform in this PR changes the problem from an object allocation to string concatenation one. The latter is around 7-20x faster depending on the amount of static bits inside a template. For the dynamic parts, this transform falls back to the `automatic` runtime one internally.

## Quick JSX history

Traditionally, JSX leaned quite heavily on the Garbage Collector because of its heritage as a templating engine for in-browser rendering.

```tsx
// input
const a = <div className="foo" tabIndex={-1}>hello<span /></div>

// classic Transform
const a = React.createElement(
  "div",
  { className: "foo", tabIndex: -1 },
  "hello",
  React.createElement("span", null)
);
```

Every element expands to two object allocations, a number which grows very quickly the more elements you have.

```tsx
const a = {
  type: "div",
  props: {
    className: "foo",
    tabIndex: -1,
    children: [
      "hello",
      {
        type: "span",
        props: null
      }
    ]
  }
}
```

The later introduced `automatic` runtime transform, didn't change that fact and merely brought auto importing the JSX factory function, some special handling of the `key` prop to the table and putting `children` directly into the props object. The main goal of that transform was to improve the developer experience and avoid an internal deopt that most frameworks ran into by having to copy the `children` argument around.

```tsx
// input
const a = <div className="foo" tabIndex={-1}>hello<span /></div>

// automatic Transform
import { jsx } from "react/jsx-runtime";
const a = jsx(
  "div",
  {
    className: "foo",
    tabIndex: -1,
    children: [
      "hello",
      jsx("span", null)
    ]
  }
})
```

...still expands to many objects:

```jsx
const a = {
  type: "div",
  props: {
    className: "foo",
    tabIndex: -1,
    children: [
      "hello",
      {
        type: "span",
        props: null
      }
    ]
  }
}
```

Both transforms are quite allocation heavy and cause a lot of GC pressure. JSX element is at minimum converted into two object allocations and for many frameworks even a third one if they are backed by fiber nodes. This easily leads to +3000 allocations per request.

## Precompiling JSX at transpile time

The proposed transform in this PR moves all that work to transpilation time and pre-serializes all the static bits of the JSX template.

```tsx
// input
const a = <div className="foo" tabIndex={-1}>hello<span /></div>

// this PR
import { jsxssr, jsxattr } from "react/jsx-runtime";

const tpl = ['<div class="foo" ', '>hello<span></span></div>']
const a = jsxssr(tpl, jsxattr('tabindex', -1));
```

The `jsxssr` function can be thought of as a tagged template literal. It has very similar semantics around concatenating an array of static strings with dynamic content. To note here is that the `className` has been automatically transformed to `class`, same for `tabIndex` -> `tabindex`. Instead of creating thousands of short lived objects, the `jsxssr` function essentially just calls `Array.join()` conceptually. Only the output string is allocation, which makes this very fast.

## Benchmarks

I've put this transform through the tests in a benchmark. The first is the `automatic` transform with Preact, second is the transform from this PR rendered by Preact and third is a custom HTML transform that skips component instances entirely. The latter is obviously the fastest, as component instances are quite heavy too, and currently those need to be instantiated for backwards compatibility reasons in Preact.

- precompiled Preact: **~7.5x faster**
- fully precompiled HTML: **~20x faster**

Screenshot 2023-10-20 at 20 26 34

## Usefulness for the ecosystem

The transform in this PR was intentionally designed in a way that makes it independent of any framework. It's not tied to Preact or Fresh at all. Rather, by setting a custom `jsxImportSource` config in `deno.json` you can point it to your own factory functions. Each factory function needs to implement the following functions:

- `jsx(type: Function, props: Record<string, unknown>, key?: unknown)` the `automatic` runtime jsx factory for dynamic elements or components.
- `jsxssr(tpl: string[], ...dynamicParts: unknown[])` the main templating function which merges the static bits and with the dynamic bits
- `jsxattr(name: string, value: unknown)` used to serialize dynamic attributes when an element can be serialized. By using a function call frameworks can also return an empty string for event listeners for example and avoid them being serialized

Note: That custom implementations don't have to adhere to specific return values. The transform only expects a certain signature, so if you can also use it for frameworks which don't need to materialize component instances.

If you're interested in more of the transpilation output, checkout the tests.

### CODE REVIEW

**Comment by marvinhagemeister** [comment]
I think the overall JSX transform is now ready to be explored in the real apps. Only thing left is adding more attribute name serialization mappings, and probably some rust stuff as I just hacked to get it to work. Other than that I think I accounted for all use cases I could come up with. I think the transform is ready.

File: src/transpiling/jsx_string.rs
Comment (bartlomieju) [inline src/transpiling/jsx_string.rs:274]: This is fine 👍

File: src/transpiling/jsx_string.rs
Comment (bartlomieju) [inline src/transpiling/jsx_string.rs:267]:
```suggestion
    value: str.as_str().into(),
```

File: src/transpiling/jsx_string.rs
Comment (bartlomieju) [inline src/transpiling/jsx_string.rs:786]: let's remove this

File: src/transpiling/jsx_string.rs
Comment (bartlomieju) [inline src/transpiling/jsx_string.rs:951]: Let's remove this

File: src/transpiling/jsx_string.rs
Comment (bartlomieju) [inline src/transpiling/jsx_string.rs:18]: Let's add `JsxString::new` that will allow to configure various options on creation - `templates` or `next_index` are internal fields that callers shouldn't need to care about.

Comment (bartlomieju) [reply src/transpiling/jsx_string.rs:18]: It would also be good to test various options in the test cases

File: src/transpiling/mod.rs
Comment (bartlomieju) [inline src/transpiling/mod.rs:35]: We need to expose `jsx_string::JsxString` so that CLI can consume it

File: src/transpiling/jsx_string.rs
Comment (bartlomieju) [inline src/transpiling/jsx_string.rs:296]:
```suggestion
  pub fn new(
```

File: src/transpiling/mod.rs
Comment (bartlomieju) [inline src/transpiling/mod.rs:322]: This still needs to be tested.

File: src/transpiling/jsx_precompile.rs
Comment (dsherret) [inline src/transpiling/jsx_precompile.rs:864]: Why do an std::cmp::max here? Looks like it should just be `1 + dynamic_exprs.len()`?

File: src/transpiling/jsx_precompile.rs
Comment (dsherret) [inline src/transpiling/jsx_precompile.rs:791]: Seems this can be removed or it's a bug? If it's a bug, perhaps we should add a test?

**Review by dsherret — CHANGES_REQUESTED** [review]
LGTM! Just going to block this for a minutes because there's an swc change I want to land and patch before merging this one.

File: src/transpiling/mod.rs
Comment (marvinhagemeister) [inline src/transpiling/mod.rs:316]: The `jsx_automatic` flag is not a good indicator for development mode or not. Typically the dev mode is enabled by setting a different transform. Like `react-jsxdev` vs `react-jsx` in the case of TypeScript https://www.typescriptlang.org/docs/handbook/jsx.html#basic-usage .

Comment (bartlomieju) [reply src/transpiling/mod.rs:316]: It's a typo, it was supposed to be `jsx_development`.

**Review by dsherret — APPROVED** [review]
No longer blocked to merge.

**Review by marvinhagemeister — APPROVED** [review]
Let's a go!

**Pull request merged**
